src/devices/GrovePi/GrovePi.cs

```diff
@@ -0,0 +1,226 @@
+using System;
+using System.Buffers.Binary;
+using System.Device.Gpio;
+using System.Device.I2c;
+using System.IO;
+using System.Threading;
+
+namespace Iot.Device.GrovePiDevice
+{
+    /// <summary>
+    /// Create a GrovePi class
+    /// </summary>
+    public class GrovePi : IDisposable
+    {
+        private I2cDevice _i2cDevice;
+        private readonly bool _autoDispose;
+        private const byte MaxRetries = 4;
+
+        /// <summary>
+        /// The default GrovePi I2C address is 0x04
+        /// Other addresses can be use, see GrovePi documentation
+        /// </summary>
+        public const byte DefaultI2cAddress = 0x04;
+
+        /// <summary>
+        /// The maximum ADC, 12 bit so 1023 on GrovePi
+        /// </summary>
+        public int MaxAdc => 1023;
+        /// <summary>
+        /// Contains the GrovePi key information
+        /// </summary>
+        public Info GrovePiInfo { get; internal set; }
+
+        /// <summary>
+        /// GrovePi constructor
+        /// </summary>
+        /// <param name="i2cDevice">The I2C device. Device address is 0x04</param>
+        /// <param name="autoDispose">True to dispose the I2C device when disposing GrovePi</param>
+        public GrovePi(I2cDevice i2cDevice, bool autoDispose = true)
+        {
+            _i2cDevice = i2cDevice ?? throw new ArgumentException("I2C device can't be null");
+            _autoDispose = autoDispose;
+            GrovePiInfo = new Info() { SoftwareVersion = GetFirmwareVerion() };
+        }
+
+        public void Dispose()
+        {
+            if (_autoDispose)
+            {
+                _i2cDevice?.Dispose();
+                _i2cDevice = null;
+            }
+        }
+
+        /// <summary>
+        /// Get the firmware version
+        /// </summary>
+        /// <returns>GroovePi firmware version</returns>
+        public Version GetFirmwareVerion()
+        {
+            WriteCommand(GrovePiCommands.Version, 0, 0, 0);
+            var inArray = ReadCommand(GrovePiCommands.Version, 0);
+            return new Version(inArray[1], inArray[2], inArray[3]);
+        }
+
+        /// <summary>
+        /// Write a GrovePi command
+        /// </summary>
+        /// <param name="commands">The GrovePi command</param>
+        /// <param name="pin">The pin to write the command</param>
+        /// <param name="param1">First parameter</param>
+        /// <param name="param2">Second parameter</param>
+        public void WriteCommand(GrovePiCommands commands, GrovePort pin, byte param1, byte param2)
+        {
+            Span<byte> outArray = stackalloc byte[4] { (byte)commands, (byte)(pin), param1, param2 };
+            byte tries = 0;
+            IOException innerEx = new IOException();
+            while (tries < MaxRetries)
+            {
+                try
+                {
+                    _i2cDevice.Write(outArray);
+                    return;
+                }
+                catch (IOException ex)
```

Comment:
you can do `catch (IOException ex) when (tries < MaxRetries)`
and then `while (true)` - it will just throw on the last try - fine to leave as is too

Reply:
I4ll leave it like this so far. But I keep the pattern in mind for other cases where it may Apply better.

Reply:
Can we write something here describing why we're retrying? Ideally we don't swallow `IOException`, of course, but if there is a valid reason we should clearly state why.